### PR TITLE
Implement static pre-evaluation check, instead of `post_init` check.

### PR DESCRIPTION
**Pull request by femtomc**
Addresses #140.

### CODE REVIEW

**Comment by femtomc** [comment]
@dfm When you have some time, consider the pattern I'm introducing into the `Kernel` ABC -- instead of `evaluate`, overloads define `__evaluate__` and `evaluate` calls `__static_evaluate_check__` and then `__evaluate__`.

**Comment by dfm** [comment]
Thanks for this! I think I understand what's going on now. What do you think about instead just moving all the checks that are currently in `__post_init__` to the individual `evaluate` functions directly? It seems like it wouldn't be too painful to do that and more backwards compatible?

File: src/tinygp/kernels/stationary.py [hunk 1]
```diff
@@ -206,12 +206,16 @@ class ExpSineSquared(Stationary):
...
+        if jnp.ndim(self.scale):
+            raise ValueError(
+                "Only scalar scales are permitted for stationary kernels; use"
+                "transforms.Linear or transforms.Cholesky for more flexiblity"
+            )
         if self.gamma is None:
             raise ValueError("Missing required argument 'gamma'")
 
-    def evaluate(self, X1: JAXArray, X2: JAXArray) -> JAXArray:
-        assert self.gamma is not None
```
Comment (dfm) [inline src/tinygp/kernels/stationary.py:214]: This is in here for mypy reasons so better not to remove it, I'd say!

File: src/tinygp/kernels/stationary.py [hunk 2]
```diff
@@ -237,11 +241,15 @@ class RationalQuadratic(Stationary):
...
+        if jnp.ndim(self.scale):
+            raise ValueError(
+                "Only scalar scales are permitted for stationary kernels; use"
+                "transforms.Linear or transforms.Cholesky for more flexiblity"
+            )
         if self.alpha is None:
             raise ValueError("Missing required argument 'alpha'")
 
-    def evaluate(self, X1: JAXArray, X2: JAXArray) -> JAXArray:
-        assert self.alpha is not None
```
Comment (dfm) [inline src/tinygp/kernels/stationary.py:245]: Same as above!

**Review by dfm — CHANGES_REQUESTED** [review]

**Comment by femtomc** [comment]
Sure! That sounds great!

I was attempting to implement something which would minimize replicating calls — the problem is that there’s not a super convenient way to add a pre-check to a function which a user can define for their implementation — without doing the weird indirection in my initial PR.

I’ll change it to your suggestion!

**Pull request closed without merging**